presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.zorder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+/**
+ * The ZOrderCurveSearcher class contains functions to search for addresses within given value ranges.
+ */
+public class ZOrderCurveSearcher
+{
+    public enum ReturnType {
+        INTEGER, LONG
+    }
+
+    private final ZOrder zOrder;
+
+    private final List<Integer> encodingBits;
+    private final int totalBitLength;
+    private final int maxBitLength;
+
+    private List<ZValueRange> ranges;
+
+    private final ArrayList<ZAddressRange> addressRanges;
+
+    private ReturnType returnType;
+
+    private static int[] dimensions;
+
+    private ZOrderCurveSearcher(ZOrder zOrder, List<Integer> encodingBits, int totalBitLength, int maxBitLength)
```

Comment:
I see! Good idea. I did that previously, but it isn't good practice to have public or even protected variables in `ZOrder`, so I'm not sure how else to access these private variables (`encodingBits`, `totalBitLength`, `maxBitLength`). 
Maybe if the classes were in the same file, this would work better? But that would probably require a significant amount of refactoring.

Reply:
Another idea: would it be better if I only passed in `encodingBits`, and recalculated `totalBitLength` and `maxBitLength` from it (which takes 2 loops through `encodingBits`)?